Codeql fixes

albertony opened the pull request:
#### What is the purpose of this change?

This fixes a few CodeQL issues, reported in the project's Security page on GitHub. It would be good to at least get rid of the ones reported as `Critical` here, to avoid scaring away any potential users..

This should fix 2 of 3 `Critical` issues:
- https://github.com/rclone/rclone/security/code-scanning/129
- https://github.com/rclone/rclone/security/code-scanning/130

It should also fix 5 issues of type "Incorrect conversion between integer types" (severity `High`), all of that type except one, as well as one of 4 issues of type "Size computation for allocation may overflow" (severity `High`).

#### Was the change discussed in an issue or in the forum before?

<!--
Link issues and relevant forum posts here.
-->

#### Checklist

- [ ] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [ ] I have added tests for all changes in this PR if appropriate.
- [ ] I have added documentation for the changes if appropriate.
- [ ] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [ ] I'm done, this Pull Request is ready for review :-)

albertony commented:
The last `Critical` issue I'm not entirely sure is real, and if so how to best fix it. Any suggestions?
- https://github.com/rclone/rclone/security/code-scanning/88

albertony commented:
I also left one issue "Incorrect conversion between integer types" (severity `High`), which I actually think is false positive due to the fact that the code checks `abs > size` etc, but that CodeQL misses that fact because several 32-bit `int`s are converted to `int64` and back.. If you agree we can select "Dismiss alert" with reason "False positive" in the GitHub Security page:
- https://github.com/rclone/rclone/security/code-scanning/134

ncw commented on backend/local/metadata.go:
```diff
@@ -72,12 +72,12 @@ func (o *Object) parseMetadataInt(m fs.Metadata, key string, base int) (result i
 	if ok {
 		var err error
-		result64, err := strconv.ParseInt(value, base, 64)
+		parsed, err := strconv.ParseInt(value, base, 32)
```
I don't think this is the right fix as `int` is 64 bit on most platforms but 32 on some.

I think the right fix is probably using `strconv.Atoi` which hopefully deals with the sizes of integers properly. Looking at [the source of Atoi](https://cs.opensource.google/go/go/+/refs/tags/go1.23.0:src/strconv/atoi.go;l=242) I think it does the right thing.

albertony replied:
Right, I was too quick on this one, and thinking C++ types.

I can use `strconv.ParseInt` as well, supplying bit size `0` does the same thing - it defaults to `strconv.IntSize` and effectively returns an `int` value within the `int64` return type. It is documented that "Atoi is equivalent to ParseInt(s, 10, 0), converted to type int", however it is faster, it has  a "fast path" and then a slow path where it do call `ParseInt` internally. In this specific case, though, our function `parseMetadataInt` takes `base` as a parameter, so sending that directly into `parseInt` fits best.

ncw commented on cmd/serve/docker/options.go:
```diff
@@ -272,10 +273,22 @@ func getVFSOption(vfsOpt *vfscommon.Options, opt rc.Params, key string) (ok bool
 		intVal, err = opt.GetInt64(key)
-		vfsOpt.UID = uint32(intVal)
+		if err == nil {
+			if intVal > 0 && intVal <= math.MaxInt32 {
```
I think the comparison should be
```
if intVal >= 0 && intVal <= math.MaxUint32
```

ncw commented on cmd/serve/docker/options.go:
```diff
@@ -272,10 +273,22 @@ func getVFSOption(vfsOpt *vfscommon.Options, opt rc.Params, key string) (ok bool
 		intVal, err = opt.GetInt64(key)
-		vfsOpt.GID = uint32(intVal)
+		if err == nil {
+			if intVal > 0 && intVal <= math.MaxInt32 {
```
And this one too

ncw commented on lib/encoder/encoder.go:
```diff
@@ -190,7 +190,7 @@ func (mask *MultiEncoder) Set(in string) error {
 			out |= bits
 		} else {
-			i, err := strconv.ParseInt(part, 0, 64)
+			i, err := strconv.ParseUint(part, 0, 32)
```
This is another 32 vs 64 bit problem with uint

Not sure what the correct solution here is as there is no `strconf.Atou`

Rlone will go wrong on 32 bit platforms if there are more than 32 options though so we could make MultiEncoder a `uint32` or `uint64` though I expect that might be quite a distruptive change!

albertony replied:
`strconv.ParseUint(part, 0, 0)` does the same.

ncw commented on fs/config/obscure/obscure.go:
```diff
@@ -47,6 +48,9 @@ func crypt(out, in, iv []byte) error {
 // This is done by encrypting with AES-CTR
 func Obscure(x string) (string, error) {
 	plaintext := []byte(x)
+	if math.MaxInt32-aes.BlockSize < len(plaintext) {
```
I think if someone has put 4 GiB of text in a string we are obscuring then perhaps they deserve rclone to blow up! The fix is OK though.

ncw reviewed (commented):
I put some queries inline about these. Some of them are quite tricky!

ncw commented:
> The last `Critical` issue I'm not entirely sure is real, and if so how to best fix it. Any suggestions?
> 
> * [Uncontrolled data used in network request](/rclone/rclone/security/code-scanning/88)

The p.url comes from something received from the plex server. I remember having a discussion about this with someone. 

https://github.com/rclone/rclone/blob/705c72d2939a875978d77badad2b56f7e6eb6ca0/backend/cache/plex.go#L159-L164

That is fairly common in REST protocols though so it might just be complaining about the `fmt.Sprintf("%s%s"` which should really be some sort of `url.Join`.

> I also left one issue "Incorrect conversion between integer types" (severity `High`), which I actually think is false positive due to the fact that the code checks `abs > size` etc, but that CodeQL misses that fact because several 32-bit `int`s are converted to `int64` and back.. If you agree we can select "Dismiss alert" with reason "False positive" in the GitHub Security page:
> 
> * [Incorrect conversion between integer types](/rclone/rclone/security/code-scanning/134)

I think that is a false positive, yes and it should be dismissed. I'll leave the pressing the button to you.

albertony commented:
Amended according to your comments.

ncw reviewed (approved):
Looks great, thanks

Please merge:-)

albertony commented:
Thanks. I have verified that the results of the latest scan had the expected improvements - except one new issue popped out, because it was previously masked by one of the other issues that I fixed in this PR. Added a new PR to fix this as well, to get it out of the way: https://github.com/rclone/rclone/pull/8021
